tensorboard/uploader/proto/write_service.proto

```diff
@@ -119,6 +124,125 @@ message WriteScalarResponse {
   // This is empty on purpose.
 }
 
+// The blob sequence writing procedure works as follows:
+//  * The caller requests creation of a new BlobSequence, via the
+//    CreateBlobSequence RPC. The response provides a new BlobSequence ID.
+//  * The caller writes each blob in the sequence separately via the WriteBlob
+//    RPC as described below, providing the blob sequence ID and sequence index
+//    for each blob.
+//  * Multiple WriteBlob streams may operate in parallel, with no constraint
+//    on their ordering.
+
+// Blob sequence uploads are resumable, as follows:
+//  * The client starts the upload process with a new CreateBlobSequence RPC.
+//  * The blob sequences are uniquely keyed by experiment, run, tag, and step.
+//    Since this is a second (or later) attempt regarding an existing sequence,
+//    a new sequence is not created.  Instead the existing sequence ID is
+//    returned.
+//  * The client attempts to upload each blob in the sequence.  These uploads
+//    resume or may be skipped entirely, as needed, according to the below
+//    procedure.
+
+// The blob writing procedure is as follows.  It is intrinsically resumable.
+//  * The caller instantiates a WriteBlob RPC, which is a bidirectional stream.
+//  * The first of these requests must provide write_blob_first_request_data,
+//    and the subsequent ones must not.
+//  * The first request should send no actual data; it acts as a request for the
```

Comment:
IMO mandating that the first request be empty feels like it unduly favors an uncommon case (resuming an partial upload where the client is unsure what progress it made) over the far more common case (where the entire blob is <4MB and we could otherwise upload the entire thing in a single request).

If we make that part optional, the client could still use that to check progress in cases where it knows it's resuming an upload or if it knows that it's using a blob sequence that already exists, but in the common case where it's just churning through a time series and for each step creating 1 blob sequence and 1 blob, I think that would basically get us a free reduction from 3 to 2 round trips per step.

One other consideration too is that if we had CreateBlobSequence/GetBlobSequence return the WriteBlobRequest properties (for the blob if it already exists) then the client could still do the check against sending redundant data, but as part of the first RPC rather than needing a second intermediary RPC.

Reply:
Agreed that the requirement to send no data on the first request was too strong.  It's fine either way, so we can leave the tradeoff to the client (e.g., it may choose a blob size threshold below which it sends data optimistically, and above which it checks status first, etc.).

I did previously have the `GetOrCreateBlobSequenceResponse` report `repeated WriteBlobResponse blob_statuses;`, but removed it on the theory that it added complexity, and that the extra backend queries needed to populate that would not be worth it most of the time.  Of course the empty-request plan recapitulated those costs, only much worse.  Oops.

Anyway, in the updated plan we do neither check, so the common case is as cheap as possible.  Additional costs are borne only when resuming uploads.  (We can always reinstate `GetOrCreateBlobSequenceResponse.blob_statuses` later, so let's not block on that now).